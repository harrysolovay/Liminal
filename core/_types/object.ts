import type { Type } from "../Type.ts"
import { declareType } from "./_declareType.ts"

export function object<F extends Record<string, Type>>(
  fields: F,
): Type<{ [K in keyof F]: F[K]["T"] }> {
  return declareType(() => object, [fields])
}
