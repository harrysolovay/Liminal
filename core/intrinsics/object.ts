import { Rune } from "../Rune.ts"
import { consumeType } from "./_common.ts"

export function object<F extends Record<string, Rune>>(
  fields: F,
): Rune<{ [K in keyof F]: F[K]["T"] }, F[keyof F]["E"]> {
  return Rune({
    kind: "object",
    self: () => object,
    args: [fields],
    consume: consumeType,
  })
}
