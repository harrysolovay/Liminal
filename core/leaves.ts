import { Ty } from "./Ty.ts"

export type none<P extends string = string> = Ty<null, P>
export const none: Ty<null, never> = Ty((description) => ({
  type: "null",
  description,
}))

export type bool<P extends string = string> = Ty<boolean, P>
export const bool: Ty<boolean, never> = Ty((description) => ({
  type: "boolean",
  description,
}))

export type num<P extends string = string> = Ty<number, P>
export const num: Ty<number, never> = Ty((description) => ({
  type: "number",
  description,
}))

export type str<P extends string = string> = Ty<string, P>
export const str: Ty<string, never> = Ty((description) => ({
  type: "string",
  description,
}))
