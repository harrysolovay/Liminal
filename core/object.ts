import type { Type } from "../Type.ts"
import { declare } from "../TypeDeclaration.ts"
import type { Expand } from "../util/type_util.ts"

export function object<F extends Record<string, Type>>(
  fields: F,
): Type<Expand<NativeObject<F>>, {}, F[keyof F]["P"]> {
  const keys = Object.keys(fields)
  return declare<NativeObject<F>>()({
    name: "object",
    source: {
      factory: object,
      args: { fields },
    },
    subschema: (visit) => ({
      type: "object",
      properties: Object.fromEntries(keys.map((k) => [k, visit(fields[k]!)])),
      additionalProperties: false,
      required: keys,
    }),
    process: (value, visit) =>
      Object.fromEntries(keys.map((k) => [k, visit(value[k]!, fields[k]!, k)])) as never,
  })
}

export type NativeObject<F extends Record<string, Type>> = { [K in keyof F]: F[K]["T"] }
