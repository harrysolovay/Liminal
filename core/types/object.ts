import { type AnyType, Type } from "../Type.ts"

export function object<F extends Record<string, AnyType>>(
  fields: F,
): Type<{ [K in keyof F]: F[K]["T"] }, F[keyof F]["P"]> {
  return Type({
    name: "object",
    source: {
      factory: object,
      args: [fields],
    },
  })
}
