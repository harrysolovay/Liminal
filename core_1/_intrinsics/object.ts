import { Rune } from "../Rune.ts"
import type { Type } from "../Type/Type.ts"

export function object<F extends Record<string, Type>>(
  fields: F,
): Type<{ [K in keyof F]: F[K]["T"] }> {
  return Rune({
    kind: "Type",
    self: () => object,
    args: [fields],
  })
}
