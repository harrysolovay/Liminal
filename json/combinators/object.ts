import { type AnyType, Type } from "../../core/mod.ts"

export function object<F extends Record<string, AnyType>>(
  fields: F,
): Type<{ [K in keyof F]: F[K]["T"] }, F[keyof F]["P"]> {
  return Type({
    factory: object,
    args: [fields],
  })
}
