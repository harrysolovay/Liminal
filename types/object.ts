import { type AnyType, declareType, type Type } from "../core/mod.ts"

export function object<F extends Record<string, AnyType>>(
  fields: F,
): Type<{ [K in keyof F]: F[K]["T"] }, F[keyof F]["P"]> {
  const entries = Object.entries(fields)
  return declareType({
    name: "object",
    source: {
      factory: object,
      args: [fields],
    },
    visitValue: (value, visit) => {
      const visited = Object.fromEntries(
        entries.map(([key, type]) => [
          key,
          visit(
            value[key],
            type,
            () => (value) => {
              visited[key] = value
            },
            {
              formatValuePath: (leading) => `${leading}.${key}`,
              formatTypePath: (leading) => `${leading}.${key}`,
            },
          ),
        ]),
      )
      return visited as never
    },
  })
}
