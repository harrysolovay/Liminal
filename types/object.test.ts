import { object } from "./object.ts"
import { assertTySnapshot } from "test_util"
import { boolean, number, string } from "./leaves.ts"

Deno.test("object", async (t) => {
  await assertTySnapshot(t, object({ a: boolean, b: number, c: string }))
})
