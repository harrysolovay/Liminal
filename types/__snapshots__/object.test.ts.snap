export const snapshot = {};

snapshot[`object 1`] = `
{
  additionalProperties: false,
  description: "",
  properties: {
    a: {
      description: "",
      type: "boolean",
    },
    b: {
      description: "",
      type: "number",
    },
    c: {
      description: "",
      type: "string",
    },
  },
  required: [
    "a",
    "b",
    "c",
  ],
  type: "object",
}
`;

snapshot[`object 2`] = `
{
  additionalProperties: false,
  description: "Description one.",
  properties: {
    a: {
      description: "",
      type: "boolean",
    },
    b: {
      description: "",
      type: "number",
    },
    c: {
      description: "",
      type: "string",
    },
  },
  required: [
    "a",
    "b",
    "c",
  ],
  type: "object",
}
`;

snapshot[`object 3`] = `
{
  additionalProperties: false,
  description: "Description two. Description one.",
  properties: {
    a: {
      description: "",
      type: "boolean",
    },
    b: {
      description: "",
      type: "number",
    },
    c: {
      description: "",
      type: "string",
    },
  },
  required: [
    "a",
    "b",
    "c",
  ],
  type: "object",
}
`;
